### PR TITLE
[ROCm] Add support for SymmetricMemory

### PR DESCRIPTION
**Pull request by pragupta**
With the latest IFU into rocm6.3_internal_testing branch, we pulled in SymmetricMemory code which is being used by intra node communication also. SymmetricMemory also introduces a new memory allocator called CUDASymmetricMemoryAllocator. intra_node_comm is now using this new allocator to manage memory.

### CODE REVIEW

**Comment by pruthvistony** [comment]
@pragupta ,
The changes look good. 
Can we merge this after the upstream patch is merged.

**Comment by pruthvistony** [comment]
Lets try on MI200 and check the UTs.

**Comment by rocm-mici** [comment]
Jenkins build for [a4afe0980223350329565cc423e66bf270f61bbf](https://github.com/ROCm/pytorch/commit/a4afe0980223350329565cc423e66bf270f61bbf) commit finished as FAILURE
Links: [Blue Ocean view](http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/pytorch%2Fpytorch-ci-pipeline/detail/PR-1569/9//pipeline) / [Build artifacts](http://ml-ci-internal.amd.com:8080/job/pytorch/job/pytorch-ci-pipeline/job/PR-1569/9/artifact)

**Comment by rocm-mici** [comment]
Jenkins build for [a4afe0980223350329565cc423e66bf270f61bbf](https://github.com/ROCm/pytorch/commit/a4afe0980223350329565cc423e66bf270f61bbf) commit finished as FAILURE
Links: [Blue Ocean view](http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/pytorch%2Fpytorch-ci-pipeline/detail/PR-1569/10//pipeline) / [Build artifacts](http://ml-ci-internal.amd.com:8080/job/pytorch/job/pytorch-ci-pipeline/job/PR-1569/10/artifact)

**Comment by rocm-mici** [comment]
Jenkins build for [a4afe0980223350329565cc423e66bf270f61bbf](https://github.com/ROCm/pytorch/commit/a4afe0980223350329565cc423e66bf270f61bbf) commit finished as FAILURE
Links: [Blue Ocean view](http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/pytorch%2Fpytorch-ci-pipeline/detail/PR-1569/11//pipeline) / [Build artifacts](http://ml-ci-internal.amd.com:8080/job/pytorch/job/pytorch-ci-pipeline/job/PR-1569/11/artifact)

**Comment by rocm-mici** [comment]
Jenkins build for [a4afe0980223350329565cc423e66bf270f61bbf](https://github.com/ROCm/pytorch/commit/a4afe0980223350329565cc423e66bf270f61bbf) commit finished as FAILURE
Links: [Blue Ocean view](http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/pytorch%2Fpytorch-ci-pipeline/detail/PR-1569/12//pipeline) / [Build artifacts](http://ml-ci-internal.amd.com:8080/job/pytorch/job/pytorch-ci-pipeline/job/PR-1569/12/artifact)

**Comment by rocm-mici** [comment]
Jenkins build for [a4afe0980223350329565cc423e66bf270f61bbf](https://github.com/ROCm/pytorch/commit/a4afe0980223350329565cc423e66bf270f61bbf) commit finished as FAILURE
Links: [Blue Ocean view](http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/pytorch%2Fpytorch-ci-pipeline/detail/PR-1569/13//pipeline) / [Build artifacts](http://ml-ci-internal.amd.com:8080/job/pytorch/job/pytorch-ci-pipeline/job/PR-1569/13/artifact)

**Comment by jataylo** [comment]
> @pragupta , The changes look good. Can we merge this after the upstream patch is merged.

cc: @pruthvistony not sure if this upstreamed yet, but we need this for 6.3. Essential for gpt-fast.

**Review by pruthvistony — APPROVED** [review]

**Pull request merged**
